cdap-security/src/main/java/co/cask/cdap/security/guice/SecureStoreModules.java

```diff
@@ -66,43 +86,99 @@ public final Module getDistributedModules() {
       this.cConf = cConf;
       this.injector = injector;
     }
 
     @Override
     @SuppressWarnings("unchecked")
     public T get() {
-      if (!SecureStoreUtils.isKMSBacked(cConf)) {
-        throw new IllegalArgumentException("File based secure store is not supported in distributed mode. " +
+      boolean kmsBacked = KMS_BACKED.equalsIgnoreCase(cConf.get(Constants.Security.Store.PROVIDER));
+      if (kmsBacked && isKMSCapable()) {
+        try {
+          return (T) injector.getInstance(Class.forName("co.cask.cdap.security.store.KMSSecureStore"));
+        } catch (ClassNotFoundException e) {
+          // KMSSecureStore could not be loaded
+          throw new RuntimeException("CDAP KMS classes could not be loaded. " +
+                                       "Please verify that CDAP is correctly installed");
+        }
+      }
+      if (kmsBacked) {
+        throw new IllegalArgumentException("Could not find classes required for supporting KMS based secure store. " +
+                                             "KMS backed secure store depends on " +
+                                             "org.apache.hadoop.crypto.key.kms.KMSClientProvider being available. " +
+                                             "This is supported in Apache Hadoop 2.6.0 and up and on " +
+                                             "distribution versions that are based on Apache Hadoop 2.6.0 and up.");
+      }
+
+      if (FILE_BACKED.equalsIgnoreCase(cConf.get(Constants.Security.Store.PROVIDER))) {
+        throw new IllegalArgumentException("Only KMS based provider is supported in distributed mode. " +
                                              "To be able to use secure store in a distributed environment you" +
                                              "will need to use the Hadoop KMS based provider.");
       }
+      return (T) injector.getInstance(DummySecureStore.class);
+    }
+
+    private static boolean isKMSCapable() {
       try {
-        if (SecureStoreUtils.isKMSCapable()) {
-          return (T) injector.getInstance(Class.forName("co.cask.cdap.security.store.KMSSecureStore"));
-        }
-      } catch (ClassNotFoundException e) {
-        // KMSSecureStore could not be loaded
-        LOG.warn("Could not find classes required for supporting KMS based secure store. KMS backed secure store " +
-                   "depends on org.apache.hadoop.crypto.key.kms.KMSClientProvider being available. " +
-                   "This is supported in Apache Hadoop 2.6.0 and up and on distribution versions that are based " +
-                   "on Apache Hadoop 2.6.0 and up.");
+        // Check if required KMS classes are present.
+        Class.forName("org.apache.hadoop.crypto.key.kms.KMSClientProvider");
+        return true;
+      } catch (ClassNotFoundException ex) {
+        // KMS is not supported.
+        return false;
+      }
+    }
+  }
+
+  @Singleton
```

Comment:
No.

> On Aug 2, 2016, at 6:06 PM, Bhooshan Mogal <notifications@github.com> wrote:
> 
> In cdap-security/src/main/java/co/cask/cdap/security/guice/SecureStoreModules.java <https://github.com/caskdata/cdap/pull/6293#discussion_r73266074>:
> 
> > -        // KMSSecureStore could not be loaded
> > -        LOG.warn("Could not find classes required for supporting KMS based secure store. KMS backed secure store " +
> > -                   "depends on org.apache.hadoop.crypto.key.kms.KMSClientProvider being available. " +
> > -                   "This is supported in Apache Hadoop 2.6.0 and up and on distribution versions that are based " +
> > -                   "on Apache Hadoop 2.6.0 and up.");
> > +        // Check if required KMS classes are present.
> > +        Class.forName("org.apache.hadoop.crypto.key.kms.KMSClientProvider");
> > +        return true;
> > +      } catch (ClassNotFoundException ex) {
> > +        // KMS is not supported.
> > +        return false;
> > +      }
> > +    }
> > +  }
> > +
> > +  @Singleton
> the provider needs to be singleton?
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub <https://github.com/caskdata/cdap/pull/6293/files/c6b47822cc9ec98a5f764fa61727856d1643f30e#r73266074>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AAFGk0WtnoGHf06YQiG5c_SpElb37st8ks5qb-mvgaJpZM4Ja37Y>.
>